fmpz/mul_2exp.c

```diff
@@ -9,38 +10,103 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
-#include <gmp.h>
-#include "flint.h"
-#include "ulong_extras.h"
 #include "fmpz.h"
+#include <string.h>
 
 void
 fmpz_mul_2exp(fmpz_t f, const fmpz_t g, ulong exp)
 {
-    fmpz d = *g;
+    slong c1 = *g;
+    ulong c1abs, c1bits;
+    __mpz_struct * mf;
 
-    if (!COEFF_IS_MPZ(d))       /* g is small */
+    if (c1 == 0)
     {
-        ulong dabs = FLINT_ABS(d);
-        ulong bits = FLINT_BIT_COUNT(dabs);
-        if (bits == 0)
+        fmpz_zero(f);
+        return;
+    }
+
+    c1abs = FLINT_ABS(c1);
+    c1bits = FLINT_BIT_COUNT(c1abs);
+
+    if (c1bits + exp <= FLINT_BITS - 2)  /* Result fits inside a small fmpz */
```

Comment:
If you are going to use FLINT_BITS - 2 instead of the macro designed for this, then at least make it a constant defined in fmpz.h so it only has to be changed in one place when we make it FLINT_BITS - 3 for garbage collection purposes (to allow mark and sweep).

Reply:
There are multiple instances where `FLINT_BITS - 2` is used (I think over 100 times). But nonetheless, it's a good idea to have a macro for this. Thanks.